### ISSUE TITLE
Custom combining classes for Thai script

### ISSUE THREAD

**Issue by ohbendy** [harfbuzz/harfbuzz#4172]
Apologies if I'm misunderstanding how things work, I'm not sure of all the technicalities here.

Peter Constable's document ['Canonical ordering of marks in Thai script'](https://www.unicode.org/L2/L2018/18216-thai-order.pdf) shows us some unusual combinations required for minority orthographies that use Latin diacritics such as the tilde and macronbelow. The macronbelow has the combining class 220 and the tilde has 230, which means they get normalised after Thai combining vowels. As Peter notes, this doesn't give the correct visual appearance, because in actual usage the macronbelow should appear between a base consonant and a belowmark (CC=103), and the tilde should appear before a tonemark (CC=107). 

Peter notes 'Thai rendering engines can be designed to display the left-hand [correct] result in each case.' In a [follow-up document](https://www.unicode.org/L2/L2018/18248-marks-for-thai.pdf), Richard Wordingham proposes using a custom order for combining classes in Thai: 'the rearrangements can be given effect by selecting bespoke canonically representative
strings not with the mark class order <9, 103, 107, 220, 230>, but with the mark class order <220, 9, 103,
230, 107>, and rendering that.'

We can test the two problematic orderings with the sequences 0E01 0303 0E48 ก่̃ (the tilde should sit on the consonant and the tonemark on the tilde) and 0E01 0331 0E39 กู̱  (where the macronbelow should sit below the consonant with the vowel beneath it).

One other problem occurs in the sequence 0E01 0E3A 0E39 กฺู. Here the canonical order matches the desired visual appearance, but Harfbuzz seems to be incorrectly reordering the 0E3A after the vowel. We want the consonant to carry the 0E3A and then the vowel to sit beneath it.

**Comment by dscorbett** [comment]
The reordering of U+0E3A is intentional:
https://github.com/harfbuzz/harfbuzz/blob/79233a149209e3da199bb4e2f74271668502c574/src/hb-ot-shaper-thai.cc#L293-L298

See also #1088.

**Comment by ohbendy** [comment]
Thanks. Is there any explanation of the rationale for this reordering? It doesn't seem beneficial to put things in an order which isn't visually correct.

Note Lao now has 0EBA corresponding to Thai 0E3A. That renders correctly, please don't change it :)

**Comment by ohbendy** [comment]
I just checked on Windows 10, and it is doing the (incorrect) reordering of 0E3A after 0E38 and 0E39. I can't understand why it would have a purposeful override of the combining classes to render things incorrectly. Does Harfbuzz have to inherit the problem?

**Comment by behdad** [comment]
> Does Harfbuzz have to inherit the problem?

Not if it's definitely wrong.

**Comment by ohbendy** [comment]
As Peter's document explains on p6, the phinthu needs to be closer to the consonant than the vowels:

![Screenshot 2023-03-28 at 19 31 18](https://user-images.githubusercontent.com/12471463/228334186-93d68f14-fd63-4e14-b1c0-24576f15696c.png)

I'd say that makes the current behaviour definitely wrong, but I'd still like to understand what must have been a purposeful decision to reverse that order.

**Comment by behdad** [comment]
Yeah it looks like we need a much more involved Thai shaper to implement that proposal.

**Comment by ohbendy** [comment]
I'll do the re-reordering in the fonts for now, and will follow this discussion for updates. Richard Wordingham's implementation looks simpler in that second document. I'm happy to answer questions where I can :)

**Comment by behdad** [comment]
@dscorbett Is this something you can I can look into?

**Comment by dscorbett** [comment]
How much do we care about matching Uniscribe? If HarfBuzz behaves differently from Uniscribe, and Thai fonts start relying on the shaping engine and stop implementing reordering themselves, people using HarfBuzz will produce text that looks right to them but wrong on Windows. I am wary of introducing incompatibilities except when Uniscribe is egregiously wrong, like when it adds dotted circles to attested clusters, but that is not the case here. The current behavior is unhelpful, but it isn’t wrong: Unicode doesn’t specify the proper rendering of these sequences and you can use U+034F to induce whatever rendering you want. However, if no one is actually working around the problem with U+034F (or U+200C or similar), then the practical reality is the same as if Uniscribe were wrong; it that case it could make sense to fix the problem in HarfBuzz.

I don’t know why Uniscribe reorders U+0E3A (BELOW2) after U+0E38 and U+0E39 (BELOW1). The intent must have been to allow them to be used together, or else they be in the same diacritic class, but it’s not clear why it goes out of its way to use this unhelpful order. Maybe it was inspired by scripts like Malayalam where the vowel killer is also a vowel modifier and should therefore follow vowel signs.

If we do change anything, the best solution is to modify the CCC values 103 and 107 to follow 220 and 230. What to do about U+0E3A THAI CHARACTER PHINTHU is less clear. [L2/18-216](https://www.unicode.org/L2/L2018/18216-thai-order.pdf) and [L2/18-248](https://www.unicode.org/L2/L2018/18248-marks-for-thai.pdf) recommend different orders for <U+0E3A, U+0331>. The former recommends leaving them in NFC order:

> assuming no unforeseen requirements of other orthographies, the preferred appearance can be chosen solely on the basis of consistency with the canonical ordering.

The latter recommends the opposite order:

>Note that preferred form treats U+0331 as being more likely to be a nukta, its main use in the Thai script, as used to give English pronunciation in some English-Thai dictionaries, and allows for U+0E3A to be used as a vowel killer, its general use when giving pronunciations, and thus potentially applicable to consonants with a nukta.
